samples/kofu-coroutines-validation/src/main/kotlin/com/sample/Handlers.kt

```diff
@@ -0,0 +1,21 @@
+package com.sample
+
+import kotlinx.coroutines.FlowPreview
+import kotlinx.coroutines.reactive.awaitSingle
+import org.springframework.web.reactive.function.server.ServerRequest
+import org.springframework.web.reactive.function.server.ServerResponse
+import org.springframework.web.reactive.function.server.ServerResponse.badRequest
+import org.springframework.web.reactive.function.server.ServerResponse.ok
+import org.springframework.web.reactive.function.server.awaitBody
+
+@FlowPreview
+@Suppress("UNUSED_PARAMETER")
+class UserHandler {
+
+    suspend fun createApi(request: ServerRequest): ServerResponse =
+            request.awaitBody<User>()
+                    .validate()
+                    .leftMap { mapOf("details" to it.details()) }
+                    .fold(badRequest()::syncBody, ok()::syncBody)
```

Comment:
I'm wondering how I can improve this line to something like `fold(badRequest()::bodyAndAwait, ok()::bodyAndAwait)` instead of `awaitSingle`.

`suspension function can be called only within coroutine body` shows.

Reply:
I think I would add a Kotlin extension that would take suspending functions parameters. I have created https://github.com/making/yavi/issues/14 related issue. On my side, I have polished coroutines extensions to make that use case possible, see https://github.com/spring-projects/spring-framework/issues/22899. Our Spring Framework CI is currently broken so make sure https://build.spring.io/browse/SPR-PUB is green again to get the related changes.